test/puppetlabs/ssl_utils/core_test.clj

```diff
@@ -500,6 +502,59 @@
             (is (revoked? updated-crl cert))
             (is (revoked? updated-crl cert2))))))))
 
+(deftest validate-cert-chain-and-crls-test
+ (testing "validate cert chain against CRLs"
+   (let [single-cert-chain (-> "crl_validation/1-cert-chain.pem" open-ssl-file pem->certs)
+         single-crl-chain (-> "crl_validation/1-crl-chain.pem" open-ssl-file pem->crls)
+         cert-chain (-> "crl_validation/3-cert-chain.pem" open-ssl-file pem->certs)
+         crl-chain (-> "crl_validation/3-crl-chain.pem" open-ssl-file pem->crls)
+         other-crl-chain (-> "crl_validation/other-3-crl-chain.pem" open-ssl-file pem->crls)
+         crl-chain-missing-crls (-> "crl_validation/2-crl-chain.pem" open-ssl-file pem->crls)
+         crl-chain-extra-crls (-> "crl_validation/4-crl-chain.pem" open-ssl-file pem->crls)
+         cert-with-expired-crl (-> "crl_validation/cert-with-expired-crl.pem" open-ssl-file pem->certs)
+         expired-crl (-> "crl_validation/expired-crl.pem" open-ssl-file pem->crls)
+         cert-with-not-valid-crl (-> "crl_validation/cert-with-not-valid-crl.pem" open-ssl-file pem->certs)
+         not-yet-valid-crl (-> "crl_validation/not-yet-valid-crl.pem" open-ssl-file pem->crls)
+         cert-chain-with-crl-bad-sig (-> "crl_validation/cert-with-crl-bad-sig.pem" open-ssl-file pem->certs)
+         crl-chain-bad-signature (-> "crl_validation/crl-with-bad-signature.pem" open-ssl-file pem->crls)
+         cert-chain-with-revoked-cert (-> "crl_validation/cert-chain-with-revoked-cert.pem" open-ssl-file pem->certs)
+         crl-chain-cert-revoked (-> "crl_validation/crl-chain-with-cert-revoked.pem" open-ssl-file pem->crls)
+         missing-crls-error-message (if (SSLUtils/isFIPS) #"No CRLs found for issuer*"
+                                                          #"Could not determine revocation status")
+         invalid-crls-error-message (if (SSLUtils/isFIPS) #"Cannot verify CRL."
+                                                          #"Could not determine revocation status")
+         revoked-cert-error-message (if (SSLUtils/isFIPS) #"Certificate revocation after*"
+                                                          #"Certificate has been revoked*")]
+     (testing "successful with single-cert chain and CRL"
+       (is (nil? (validate-cert-chain single-cert-chain single-crl-chain))))
+     (testing "successful when all certs have an associated CRL"
+       (is (nil? (validate-cert-chain cert-chain crl-chain))))
+     (testing "successful with extra CRLs"
+       (is (nil? (validate-cert-chain cert-chain crl-chain-extra-crls))))
+     (testing "fails with missing CRLs"
```

Comment:
mmmm okay so this doesn't actually fail if just the leaf CRL is missing.. trying to decide if I think that behavior belongs in this function or not, since, as-is, it could be used to verify any cert chain, as opposed to a CA chain specifically. kinda leaning towards not.. Thoughts?

Reply:
I think that case makes sense, since we don't actually have to check that CRL for anything. If your chain had a non-CA cert as the leaf, and the leaf CRL was missing, I assume it _would_ break. I'm indifferent to whether we test that here.

Reply:
Since the leaf CRL comes from _our_ intermediate CA, as opposed to coming from users, I'm not so worried about it being missing.

Reply:
This is a general-purpose function, but all of the testing here is around the cases where every element of the chain is a CA cert. In that context, it makes sense to not test the leaf CRL, since it will never be relevant to validating chains like that.

Reply:
right. I think I've had iterations of the `missing-crls` file where the root CRL was missing and where an intermediate CRL was missing and they both fail as we expect. It's just the leaf CRL that it doesn't seem to care about, which makes sense for the reasons you said.